iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/ReplaceThreeWindingsTransformersBy3TwoWindingsTransformers.java

```diff
@@ -0,0 +1,344 @@
+        TwoWindingsTransformerAdder t2wAdder = starVoltageLevel.getSubstation().orElseThrow()
+                .newTwoWindingsTransformer()
+                .setId(t3w.getId() + "-" + legTag)
+                .setName(t3w.getNameOrId() + "-" + legTag)
+                .setRatedU1(leg.getRatedU())
+                .setRatedU2(starVoltageLevel.getNominalV())
+                .setR(leg.getR())
+                .setX(leg.getX())
+                .setG(leg.getG())
+                .setB(leg.getB())
+                .setRatedS(leg.getRatedS())
+                .setVoltageLevel1(leg.getTerminal().getVoltageLevel().getId())
+                .setVoltageLevel2(starVoltageLevel.getId());
+
+        connect(t2wAdder, getConnectivityLegAfterCreatingInternalConnection(leg), connectivityStar);
+        TwoWindingsTransformer t2w = t2wAdder.add();
+
+        leg.getOptionalRatioTapChanger().ifPresent(rtc -> copyAndAddRatioTapChanger(t2w.newRatioTapChanger(), rtc));
+        leg.getOptionalPhaseTapChanger().ifPresent(rtc -> copyAndAddPhaseTapChanger(t2w.newPhaseTapChanger(), rtc));
+        leg.getOperationalLimitsGroups().forEach(operationalLimitGroup -> copyOperationalLimitsGroup(t2w.newOperationalLimitsGroup1(operationalLimitGroup.getId()), operationalLimitGroup));
+
+        return t2w;
+    }
+
+    private static void connect(TwoWindingsTransformerAdder t2wAdder, ConnectivityR connectivityEnd1, ConnectivityR connectivityEnd2) {
+        if (connectivityEnd1.bus == null && connectivityEnd1.connectableBus == null) {
+            t2wAdder.setNode1(connectivityEnd1.node);
+        } else {
+            t2wAdder.setConnectableBus1(connectivityEnd1.connectableBus.getId());
+            if (connectivityEnd1.bus != null) {
+                t2wAdder.setBus1(connectivityEnd1.bus.getId());
+            }
+        }
+
+        if (connectivityEnd2.bus == null && connectivityEnd2.connectableBus == null) {
+            t2wAdder.setNode2(connectivityEnd2.node);
+        } else {
+            t2wAdder.setConnectableBus2(connectivityEnd2.connectableBus.getId());
+            if (connectivityEnd2.bus != null) {
+                t2wAdder.setBus2(connectivityEnd2.bus.getId());
+            }
+        }
+    }
+
+    private static ConnectivityR getConnectivityLegAfterCreatingInternalConnection(ThreeWindingsTransformer.Leg leg) {
+        if (leg.getTerminal().getVoltageLevel().getTopologyKind().equals(TopologyKind.NODE_BREAKER)) {
+            int newNode = leg.getTerminal().getVoltageLevel().getNodeBreakerView().getMaximumNodeIndex() + 1;
+            leg.getTerminal().getVoltageLevel().getNodeBreakerView()
+                    .newInternalConnection()
+                    .setNode1(leg.getTerminal().getNodeBreakerView().getNode())
+                    .setNode2(newNode).add();
+            return new ConnectivityR(newNode, null, null);
```

Comment:
If we tried to do after removing the t3w we would need a temporary copy of almost all the attributes of the transformer, tapChangers, operationalLimits, regulating controls.

Reply:
Can't you connect it to a fictitious node, like min(existing nodes in that voltage level) - 1, and after the removal of the t3w connect it to the t3w leg node?

Reply:
The node cannot be changed once the transformer has been created

Reply:
Thank you for moving the newReportNode to the ModificationReports class. Could you add line breaks though to make it more consistent with the rest of the report methods and add a severity (Info I guess)?

``` java
return reportNode.newReportNode()
                .withMessageTemplate("replaced-t3w-by-3t2w", "Replaced ThreeWindingsTransformer by 3 TwoWindingsTransformers")
                .withSeverity(TypedValue.INFO_SEVERITY)
                .add();
```

thanks!

Reply:
Are you sure? There is the Terminal.moveConnectable method that I believe could do that.

Reply:
You are right, the node can be changed using the moveConnectable method. Doing in this way implies:

- Terminal data (active power, reactive power, ...)  of the terminals associated to the transformers that are replaced must be temporarily recorded.
- The following methods `RegulatedTerminalControllers.replaceRegulatedTerminal, replaceRegulatedTerminal, and copyTerminalActiveAndReactivePower` must be run after the new transformers have the right node and the final terminal.
- The method `getRegulatedTerminalMapping` must be modified since we will not have at the same time the replaced transformers and the new ones.
- The method `RegulatedTerminalControllers.replaceRegulatedTerminal` must be modified avoiding the access to the replaced transformers. (The regulatingTerminal of the replaced transformer can be one of their terminals)

Let me know if you prefer doing the replacement without creating the internalConnections to do all these changes.